File: v3/python/add-asset-groups/add-asset-groups.py
#!/usr/bin/env python3
#
#  Description: Script that uses the Nexpose APIv3 to create a Dynamic Asset Group based off the criteria provided.

import json
from getpass import getpass

import urllib3
import requests
from requests.auth import HTTPBasicAuth

# Prompts to collect nexpose console and login information
host = input("Enter the server name (host) for InsightVM: ")
user = input("Enter your username: ")
password = getpass("Enter your password: ")

#Function creating the asset group and printing results
def create_ag():
    """creates a dynamic asset group based off criteria and prints out the id with a url."""
    urllib3.disable_warnings()
    url = f"https://{host}:3780/api/3/asset_groups"
    headers = {
        "Content-Type": "application/json",
    }
    payload = json.dumps(
        {
            "description": "Assets with unacceptable high risk required immediate remediation.",
            "name": "High Risk Assets",
            "searchCriteria": {
                "filters": [
                    {
                        "field": "risk-score",
                        "lower": "",
                        "operator": "is-greater-than",
                        "upper": "",
                        "value": 25000,
                        "values": ["string"],
                    }
                ],
                "match": "all",
            },
            "type": "dynamic",
            "vulnerabilities": {},
        }
    )
    response = requests.post(
        url,
        auth=HTTPBasicAuth(user, password),
        headers=headers,
        data=payload,
        verify=False,
    ).json()
    agid = response["id"]
    print(
        f"Asset Group {agid} created and can be found at https://{host}:3780/group.jsp?groupid={agid}"
    )
    return


# Main
create_ag()

File: v3/ruby/add-asset-groups/add-asset-groups.rb
#!/usr/bin/env ruby
#
#                  Description : Create asset groups from a list.
#                                Filter is hard coded, but this code 
#                                can be used as a template to build asset groups with various filters.
#                                based on individual needs..
require 'httparty'
require 'csv'
require 'json'
require 'highline/import'

# Defaults: Change to suit your environment.
default_name = 'username'
default_con = '<console_ip_address>'
#
nxconsole = ask('Enter your console IP:  ') { |q| q.default = default_con }
user = ask('Enter your username:  ') { |q| q.default = default_name }
pass = ask('Enter your password:  ') { |q| q.echo = '*' }
#


#Timeout setting below is for submitting commands that run a long time , load content for example
#command will still run but throws a ruby error if the timout is exceeded.

def create_ag(name,type,nxconsole,user,pass)
   cmd = "{
		    \"description\": \"asset group called #{name}\",
		        \"name\": \"#{name}\",
			\"searchCriteria\" : {
 		          \"match\" : \"all\",
		          \"filters\" : [ {
		            \"field\" : \"operating-system\",
		            \"operator\" : \"contains\",
			    \"value\" : \"mac\"
	 } ]
	},
			    \"type\": \"#{type}\"
	
}"
#url = "https://#{nxconsole}:3780/api/3/asset_groups"
auth = {:username => user, :password => pass,:verify => false}
options = { :basic_auth => auth,
	    :headers => { 'Content-Type' => 'application/json'},
	    :verify => false,
	    :timeout => 900,
	    :body =>  cmd
	  }

results = HTTParty.post("https://#{nxconsole}:3780/api/3/asset_groups", options)
parsed = JSON.parse(results.read_body)
return parsed
end
csv_text = File.read("aglist.txt")
csv = CSV.parse(csv_text, :headers => true)
csv.each do |row|
    n = "#{row['AG Name']}"
    t = "#{row['Type']}"
    puts "Creating asset group #{n} of type #{t}"
  res = create_ag(n,t,nxconsole,user,pass)
  pp res
end
